gemrb/core/Scriptable/Scriptable.cpp

```diff
@@ -2181,11 +2183,20 @@ void Movable::WalkTo(const Point &Des, int distance)
 		return;
 	}
 
+	Point smp_pos = Point(Pos.x/16, Pos.y/12);
+	Point smp_des = Point(Des.x/16, Des.y/12);
+	int hdist = int(Distance(smp_pos, smp_des) * 1.3);
 	if (BlocksSearchMap()) area->ClearSearchMapFor(this);
-	PathListNode* newPath = area->FindPath(Pos, Des, circleSize, distance, PF_SIGHT | PF_ACTORS_ARE_BLOCKING, actor);
-	if (!newPath && actor && actor->ValidTarget(GA_CAN_BUMP)) {
+	PathListNode * newPath;
+	int dist;
+	std::tie(newPath, dist) = area->FindPath(Pos, Des, circleSize, distance, PF_SIGHT | PF_ACTORS_ARE_BLOCKING, actor);
+	// Use a simple heuristic to identify unusually long paths by comparing to the shortest geometrical path -> try pathfinding without actors blocking.
+	// Following the two sides of an rectangle with 1:2 aspect ratio is around 1.3 times longer then the diagional.
+	// Ignore very short paths (below 20) as it would lead to excesive bumping.
+	// Ignore very long paths as they could be complicated and blocking by nearby actors will not matter anyway.
```

Comment:
Good test cases include the bg1 undercity maze and the ruins in ar3900. Both have narrow walls that can make you cry. And whole party movement in general, since it appears you've tested only with a single actor.

Reply:
The bg2 starting area is not really a maze-type test, since everything is spaced much better. Still, testing the pathing from the cage area to the Ellesime room is a good way to see how it behaves on long paths and how it responds to changes (eg. closing doors).

You can just teleport to the areas. In GemRB there's a shorthand mta("ar0000") and in the originals you have MoveToArea: https://gibberlings3.github.io/iesdp/appendices/clua/bg1.htm